### DISCUSSION TITLE
Delayed service calls and/or single-use automations

### DISCUSSION THREAD

**Discussion by catshapednoodles** [frenck/spook#572] [Ideas]
It would be cool if Spook could delay/schedule service calls and/or could create single-use automations. This could be useful for manual service calls that you do so rarely that it doesn't make sense to make an automation for it, or things you only do once.

I sometimes find myself wanting to do a service call in the future, but then needing to remind myself to actually do the thing I wanted to do. It would be great if Home Assistant could do the thing for me instead.

For instance, I usually only want to update HA Core or HA OS after an hour passes. If I update in the ~15 minutes before an hour chances are that my HA instance does not get back up before the new hour starts and that causes my Energy Dashboard to be slightly incorrect. Another example is the robot vacuum, sometimes I want to start it in an hour or so, but not right now. The same with switches, lights, anything you can think of. A third example happened on 31 December. I wanted to trigger some lights at midnight for NYE, but I had to make an automation for it that I only used once and that I deleted right after because I'll probably want something else next year. For these reasons I wish something like a delayed service call or single-use automations existed.

**Comment by frenck** [comment]
I'm not understanding. If you need to delay an automation, you can do so... automations support delays 🤷

**Reply by catshapednoodles** [reply]
Right, I'm not necessarily talking about delaying an automation. My idea would be more useful for things you do in a dashboard. For example, if I tap a light in a dashboard in the HA app there's no way to turn it off in 1 hour. You can only turn it on/off right now, but not with a delay. Same goes for all entities. Unless you have a specific script for exactly this, but my idea would be most useful for thing you'd only do once. You probably don't have a script/automation for that, because you've never needed this exact delayed service call before and probably never will again.

If we had access to single-use automations, I think this could be done, but there's no way yet to dynamically create and delete automations. I think you'd need two service calls to be able to easily create single-use automations. `automation.create_from_blueprint` could for example dynamically create an automation based on a predetermined blueprint (with some variables), and `automation.delete` could for example delete an automation, which would be especially useful if it could self-reference the automation it's currently running from.

Another way I can think of is creating a new service call, called something like `service.trigger_with_delay` that simply takes a delay value and a service call. You could use that to make a delayed service call without the need to specifically create an automation for this.

Hopefully this makes more sense now :)

**Reply by spali** [reply]
I think, I have a similar use-case or even the same.
i.e. Turn off the cameras for a specific amount of time. This time should be configurable. As I don't want to rely on a fixed time plan.
I think this can be solved by a script with entity/device and duration input fields. But service would be needed to be hard-coded. Maybe this can be solved somehow with a "service selector" and your idea of a `service.trigger_with_delay`.
Or turn off the TV in 1h (depending on the negotiating skills of your child 🤣).

Independent of the perfect solution, for the service specific script variant, is there a possibility to delay something that survives a reboot?
I know we have timer helpers which can be set to restore the state. But then I would need to dynamically create a timer helper during script runtime or I would be limited to a timer per script.
Maybe dynamically creating helper entities could be a feature for spook 😉
